### ISSUE TITLE
Append commit hash to -v flag

### ISSUE THREAD

**Issue by stefanos82** [LuaJIT/LuaJIT#711]
Greetings @MikePall, I hope you are doing well.

Is it possible to append the commit hash next to `LuaJIT 2.1.0-beta3`?

Every time I compile the code I don't know which version I have compiled and in case of a bug, I don't know how to report which version is producing it.

TinyCC added it and made things clearer: `tcc version 0.9.27 - 0378168 (x86_64 Linux)`

**Comment by MikePall** [comment]
Yes, after some bigger pending changes are done, I will switch to the `$version-$hash` format.

**Comment by stefanos82** [comment]
Excellent! Should I close this ticket or leave it open until the implementation takes place?

**Comment by MikePall** [comment]
I'll close it when it's done.

**Comment by stefanos82** [comment]
Nice! Thank you @MikePall, keep up the good work :+1:

**Comment by MikePall** [comment]
Replaced by switching to rolling releases. The commit timestamp is used for the third semver number.

**Comment by clason** [comment]
> Replaced by switching to rolling releases. The commit timestamp is used for the third semver number.

This is a welcome improvement (thanks!) but still has a few minor issues from a downstream point of view -- please take the following as constructive suggestions, not complaints:

1. Since the commit hash is non-monotonic, standard functions for comparing versions will fail, assuming that this is an actual (monotonic) patch version as semver specifies. I believe the suggested way of adding commit hashes is `<version>-prerelease+<commit hash>`. (But as LuaJIT uses a rolling release, you probably want to omit `-prerelease`, so just `2.1+<commit hash>`.)
An alternative would be [hashver](https://miniscruff.github.io/hashver/), in this case `2.1.<utf date>+<commit hash>` (as in `2.1.20230822+<hash>`); this would make the patch version behave according to semver and allow software to have at least a rough check whether a found LuaJIT version is too old.

2. The current `makefile` fails when trying to install from a tarball provided by Github (as we do in Neovim); it reports a warning that `git` is missing (which is not true in this case) and creates a `luajit-2.1.` (note the trailing dot) binary; I believe absence of a `.git` directory should be handled explicitly, creating a `luajit-2.1.ROLLING` as in the header file.

3. As the built (from `git`) binary now has a constantly changing name, this makes it difficult to use in our build scripts (which we do for generating some source files). I know it's not linked to `luajit` deliberately, but maybe a compromise would be to link it to `luajit-2.1` by default (which can then be linked to `luajit` if desired)? Then the name would be constant and could be relied on.

**Comment by daurnimator** [comment]
> 2\. The current `makefile` fails when trying to install from a tarball provided by Github (as we do in Neovim); it reports a warning that `git` is missing (which is not true in this case) and creates a `luajit-2.1.` (note the trailing dot) binary; I believe absence of a `.git` directory should be handled explicitly, creating a `luajit-2.1.ROLLING` as in the header file.

I noticed in the makefile that luajit will read the version out of a file `.relver` when you don't have a git checkout (https://github.com/LuaJIT/LuaJIT/blob/54ef81f864e18729f6dafb6b1c9f9176fcb67bb5/src/Makefile#L478)
I made use of that in the Archlinux packaging of LuaJIT: https://gitlab.archlinux.org/archlinux/packaging/packages/luajit/-/blob/69e79f866a8d99217daac6a8578fe809cc9cdb86/PKGBUILD#L28

**Comment by MikePall** [comment]
1. I said it uses the *timestamp*, not the commit hash. Timestamps are monotonic -- at least the ones I create.
2. The [download page](https://luajit.org/download.html) explicitly says: "*Do not* use pseudo-releases or tarballs created by third parties". You should be using a git checkout flow.
3. This was an oversight, [v2.1 is a production branch](https://luajit.org/status.html). I've changed the top-level `Makefile` to install the symlink. Thanks for the report!

The `.relver` file is for cases where you want the `.git` directory removed, like an intermediate tarball. The workflow is then:
```
git show -s --format=%ct >.relver
rm -rf .git
# Now create your intermediate tarball from the tree. DO NOT distribute it publicly!
```

@daurnimator Your workflow depends on always picking both the commit and the associated timestamp correctly. This is error prone. There must be a better way.

**Comment by daurnimator** [comment]
> @daurnimator Your workflow depends on always picking both the commit and the associated timestamp correctly. 

Yes. The workflow I have now is:
1. run a `git pull` locally
2. inspect the commits/diff since the last time I packaged to look for any thing that might need my attention
3. get the commit id via `git rev-parse HEAD` and paste it into the PKGBUILD file
4. get the "release version" with `git show -s --format=%ct HEAD` and paste it into the PKGBUILD file
5. run the build

> This is error prone. There must be a better way.

It's no more error prone than the previous practice we had of running `git describe -v` and taking the number of "additional commits" over the beta3 tag it gave us and copy/pasting it over to the PKGBUILD file.

The "rolling release" process you've created here is unique across all projects I've worked with. A much more normal flow with regular releases would be a better way.
